Bookmarking events and generic features

gnunicorn opened the pull request:
@kumarpalsinh25 as you wanted: the riverpod infrastructure (and a widget) for bookmarking.

This is what is does/looks like (right now).

https://github.com/user-attachments/assets/6eef6d31-438d-472f-80be-fd5bfe2ae1bc

Only implemented the toggle and for events only at the moment. Though the provider should make clear how to get a full list per type (`event`, `pins`, etc) and the toggle widget should work everywhere essentially.

kumarpalsinh25 commented on app/lib/features/bookmarks/widgets/bookmark_action.dart:
```diff
@@ -0,0 +1,31 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class BookmarkAction extends ConsumerWidget {
+  final String type;
+  final String id;
```
```suggestion
  final Bookmarker bookmarker;
```

kumarpalsinh25 commented on app/lib/features/bookmarks/widgets/bookmark_action.dart:
```diff
@@ -0,0 +1,31 @@
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final isBookmarked = ref.watch(isBookmarkedProvider((type: type, id: id)));
+
```
```suggestion
final type = bookmarker.type;
final id = bookmarker.id;
```

kumarpalsinh25 commented on app/lib/features/bookmarks/actions/bookmarking.dart:
```diff
@@ -0,0 +1,20 @@
+
+Future<bool> bookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
```
```suggestion
  required Bookmarker bookmarker;
```

kumarpalsinh25 commented on app/lib/features/bookmarks/actions/bookmarking.dart:
```diff
@@ -0,0 +1,20 @@
+
+Future<bool> unbookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
```
```suggestion
  required Bookmarker bookmarker;
```

kumarpalsinh25 commented on app/lib/features/bookmarks/actions/bookmarking.dart:
```diff
@@ -0,0 +1,20 @@
+  required id,
+}) async {
+  final manager = await ref.read(bookmarksManagerProvider.future);
+  return await manager.add(key, id);
```
```suggestion
  return await manager.add(bookmarker.type, bookmarker.id);
```

kumarpalsinh25 commented on app/lib/features/bookmarks/actions/bookmarking.dart:
```diff
@@ -0,0 +1,20 @@
+  required id,
+}) async {
+  final manager = await ref.read(bookmarksManagerProvider.future);
+  return await manager.remove(key, id);
```
```suggestion
  return await manager.remove(bookmarker.type, bookmarker.id);
```

kumarpalsinh25 commented on app/lib/features/bookmarks/providers/bookmarks_provider.dart:
```diff
@@ -0,0 +1,22 @@
+  () => BookmarksManagerNotifier(),
+);
+
+final bookmarkByKeyProvider =
```
```suggestion
final bookmarkByTypeProvider =
```

kumarpalsinh25 commented on app/lib/features/bookmarks/providers/bookmarks_provider.dart:
```diff
@@ -0,0 +1,22 @@
+});
+
+final isBookmarkedProvider =
+    StateProvider.family<bool, Bookmarker>((ref, query) {
```
```suggestion
    StateProvider.family<bool, Bookmarker>((ref, bookmarker) {
```

kumarpalsinh25 commented on app/lib/features/bookmarks/providers/bookmarks_provider.dart:
```diff
@@ -0,0 +1,22 @@
+final isBookmarkedProvider =
+    StateProvider.family<bool, Bookmarker>((ref, query) {
+  final bookmarks =
+      ref.watch(bookmarkByKeyProvider(query.type)).valueOrNull ?? [];
+  return bookmarks.contains(query.id);
```
```suggestion
      ref.watch(bookmarkByKeyProvider(bookmarker.type)).valueOrNull ?? [];
  return bookmarks.contains(bookmarker.id);
```

kumarpalsinh25 reviewed (approved):
Thanks @gnunicorn for this PR. Code looks great!
I will do the needful for listing part and other module by referencing this code.

Added some minor word suggestions that doesn't stoping this PR to merge!

gnunicorn commented:
@kumarpalsinh25 I've updated the code slightly to your requests but also replaced it with a typed system. I'll merge as-is but you might still want to take a second look to understand how to use this.

kumarpalsinh25 commented on app/lib/features/bookmarks/types.dart:
```diff
@@ -1 +1,22 @@
+typedef Bookmarker = ({BookmarkType type, String id});
+
+// referencing the inner type
+enum BookmarkType {
```
Very Nice!!
